design/design-generic-references.md

```diff
@@ -0,0 +1,314 @@
+following properties:
+
+* The referred object kind and the source field cannot be tuned/changed at 
+  runtime for a given object
+* The referencer field may not always target a single referred kind. For
+  example, a URL field in Route53 could get populated by a URL of an S3 bucket
+  or access endpoint of an RDS Instance
+* The referencer field may not always target a single field in the referred
+  object
+* The value that is about to be assigned to the field require some sort of
+  transformation
+
+Some of the above issues can be solved by defining a composition. However, there
+are cases where one would like to use and refer a resource not bound by a 
+composite instance.
+
+__**NOTE**__: there is 
+[a previous work on this topic](https://github.com/crossplane/crossplane/pull/2385).
+
+## Goals
+
+The generic cross-reference should be able to:
+
+* be configurable at the runtime
+* use the existing [Kuberentes API object referemce] convention 
+  as much as possible
+* provide alternative convention in cases where [Kuberentes API object referemce]
+  convention is not suitable
+* be implemented without or light changes in RBAC
+* support referring multiple objects (namespaced and cluster-scoped) and their
+  fields
+* support value transformation before assignment
+
+## Proposal
+
+We would like to introduce a new cluster-scoped `Referable` (alternative name
+could be `InjectableValue`) type whose instances declare values that can be
+assigned to object field:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc
+      fieldPath: status.atProvider.id
+```
+
+The syntax of `fieldPath` field is similar to what we
+use in Composition, in line with [Kubernetes API Conventions on field selection].
+
+After deploying the above instance to the cluster, its reconciliation gets 
+triggered and once the referred object exists and the field is set, 
+the referable value is emitted in the object status part:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc
+      fieldPath: status.atProvider.id
+status:
+  value: foo-vpc-id # the value found in the requested field
+  conditions:
+    - type: Ready
+      status: True
+      reason: Available
+```
+
+Now, the value can be referred in the usual way within a managed resource:
+
+```yaml
+spec:
+  vpcIdRef: referable-vpc-id
+```
+
+The reference resolver could first try to find `referable-vpc-id` `VPC` instance, 
+and fallback to `referable-vpc-id` `Referable` if the former does not exist. 
+If the instance is ready, the value is read from `status.value` and 
+set to `spec.vpcId` of the managed resource. Ofcourse, if needed, and if it 
```

Comment:
```suggestion
set to `spec.vpcId` of the managed resource. Of course, if needed, and if it 
```

Reply:
Under what circumstances might it be desired to look for the `Referable` object first?